Georeference and Globe Anchor refactoring

kring opened the pull request:
There's a lot going on here, so I'm going to open this PR as a draft, clean it up, and then mark it ready for review. More to come...

kring commented:
This should be ready to review now!

j9liu commented on Runtime/CesiumGlobeAnchor.cs:
```diff
@@ -415,78 +415,99 @@ public void SetPositionUnity(double x, double y, double z)
         /// </item>
         /// <item>
-        /// If the game object's `Transform` has changed since the last time `Start`, this method,
+        /// If the game object's `Transform` has changed since the last time `OnEnable`, this method,
```
```suggestion
        /// If the game object's `Transform` has changed since the last time `OnEnable`, or if this method
```

j9liu replied:
Actually I think I misunderstood what this comment was saying, so it really should say:
```suggestion
        /// If the game object's `Transform` has changed since the last time `OnEnable`, or if a position 
        setter was called, this method ...
```

j9liu reviewed (commented):
@kring I still have to look over the backward compatibility stuff + new UI, but just wanted to mention that I did do a pass over the new CesiumGlobeAnchor file, and it looks good to me. I had a few wording nitpicks for clarity that I'll just push a commit for

j9liu reviewed (commented):
@kring -- I gave everything a thorough look, most of it looks good! I had some minor comments / suggestions and one observation about garbage collection.

j9liu commented:
@kring I spoke too soon :( it looks like the Fly-To behavior is now broken on DynamicCamera. I'll see if I can find a fix

kring commented:
@j9liu the flight interruption is fixed. The problem was the the flight controller was setting `transform.rotation`, because that used to be the only option. But setting any property of the transform triggers the transform change detection. I changed it to set the rotation via the globe anchor instead, and now flights are working correctly. I also added a test to confirm flights work and don't get interrupted.

kring commented on Editor/CesiumInspectorGUI.cs:
```diff
@@ -0,0 +1,173 @@
+        private static GUIContent defaultYContent = new GUIContent("Y");
+        private static GUIContent defaultZContent = new GUIContent("Z");
+
+        public void Double3(string label, double3 value, Action<double3> set, string tooltip, GUIContent xContent = null, GUIContent yContent = null, GUIContent zContent = null)
```
It now puts the fields on the next line if the label doesn't fit. I think it's better, but still not great. Let me know if you have ideas.
![label-sizing](https://user-images.githubusercontent.com/924374/221758369-68522951-2691-4d2c-a950-baacb83a70b4.gif)

kring commented on Editor/CesiumGlobeAnchorEditor.cs:
```diff
@@ -7,230 +9,194 @@ namespace CesiumForUnity
-            GUILayout.Label("Position (Unity Local Coordinates)", EditorStyles.boldLabel);
+        private static readonly string rotationTooltip = CesiumEditorUtility.FormatTooltip(@"
+            The rotation of the object relative to the local East-Up-North
+            (EUP) frame centered at the object's position. In the EUP frame +X points East, +Y
```
Yeup! Fixed.

kring commented on native~/Runtime/src/CesiumGlobeAnchorImpl.cpp:
```diff
@@ -2,49 +2,196 @@
+    return GlobeAnchor(UnityTransforms::fromUnity(newLocalToGlobeFixedMatrix));
+  } else {
+    // Create an anchor at the old position and move it to the new one.
+    GlobeAnchor cppAnchor(
```
Oooh good pick up here. It's not the C++ GlobeAnchor that's the cause - that's allocated on the C++ stack, so no GC impact there. But the coroutine that detects transform changes had this code:

```cs
WaitUntil waitForChanges = new WaitUntil(() => this._lastLocalToWorld != null && !this.transform.localToWorldMatrix.Equals(this._lastLocalToWorld));
```

It looks innocent, but it allocates memory from the GC heap every frame!

The cause is the `Equals` method. It has two overloads, one that takes a `Matrix4x4` and the other that takes an `object`. But what we're passing to it, `this._lastLocalToWorld`, is a `Matrix4x4?`. That can't be implicitly converted to a `Matrix4x4`, so we end up invoking the `object` overload. Converting a value type like `Matrix4x4` or `Matrix4x4?` to an `object` requires a boxing conversion. In other words, .NET allocates a new object on the managed heap and copies the value into it. So the 80 bytes is 4x4x(sizeof(float)) for the matrix, plus 16 bytes of managed object overhead.

I fixed it by calling the `Equals` method with `this._lastLocalToWorld.Value` instead.

j9liu replied:
perfect! the GC per frame went down to zero 😄

kring commented:
Thanks for the review @j9liu! I think everything has been addressed and this is ready for another look.

j9liu commented:
@kring looks great, thanks!
